// web-push-neo/dist/encryption.mjs
//#region src/encryption.ts
/**
* Web Push Message Encryption (RFC 8291 / RFC 8188 aes128gcm)
*
* Encryption flow:
*   1. Generate an ephemeral ECDH P-256 key pair and a random 16-byte salt
*   2. Derive a shared secret via ECDH with the subscriber's p256dh public key
*   3. Derive IKM: HKDF(salt=auth_secret, ikm=shared_secret, info="WebPush: info\0" || subscriber_pub || local_pub)
*   4. Derive CEK: HKDF(salt=salt, ikm=IKM, info="Content-Encoding: aes128gcm\0") -> 16 bytes
*   5. Derive nonce: HKDF(salt=salt, ikm=IKM, info="Content-Encoding: nonce\0") -> 12 bytes
*   6. Encrypt payload with AES-128-GCM using CEK and nonce
*   7. Build aes128gcm header: [salt (16)] [record_size (4)] [keyid_len (1)] [keyid (65)]
*   8. Concatenate header + ciphertext
*/
function concat(...arrays) {
	const totalLength = arrays.reduce((sum, arr) => sum + arr.length, 0);
	const result = new Uint8Array(totalLength);
	let offset = 0;
	for (const arr of arrays) {
		result.set(arr, offset);
		offset += arr.length;
	}
	return result;
}
async function hkdfDerive(salt, ikm, info, length) {
	const key = await crypto.subtle.importKey("raw", ikm, "HKDF", false, ["deriveBits"]);
	return new Uint8Array(await crypto.subtle.deriveBits({
		name: "HKDF",
		hash: "SHA-256",
		salt,
		info
	}, key, length * 8));
}
/**
* Encrypts a payload per RFC 8291 (Web Push Message Encryption) with aes128gcm encoding.
* Returns the complete body to POST to the push service (header + ciphertext).
*/
async function encryptPayload(payload, subscriberPublicKey, authSecret) {
	const payloadBytes = typeof payload === "string" ? new TextEncoder().encode(payload) : payload;
	const salt = new Uint8Array(16);
	crypto.getRandomValues(salt);
	const localKeyPair = await crypto.subtle.generateKey({
		name: "ECDH",
		namedCurve: "P-256"
	}, true, ["deriveBits"]);
	const localPublicKeyRaw = new Uint8Array(await crypto.subtle.exportKey("raw", localKeyPair.publicKey));
	const subscriberPubKey = await crypto.subtle.importKey("raw", subscriberPublicKey, {
		name: "ECDH",
		namedCurve: "P-256"
	}, false, []);
	const ikm = await hkdfDerive(authSecret, new Uint8Array(await crypto.subtle.deriveBits({
		name: "ECDH",
		public: subscriberPubKey
	}, localKeyPair.privateKey, 256)), concat(new TextEncoder().encode("WebPush: info\0"), subscriberPublicKey, localPublicKeyRaw), 32);
	const contentEncryptionKey = await hkdfDerive(salt, ikm, new TextEncoder().encode("Content-Encoding: aes128gcm\0"), 16);
	const nonce = await hkdfDerive(salt, ikm, new TextEncoder().encode("Content-Encoding: nonce\0"), 12);
	const paddedPayload = concat(payloadBytes, new Uint8Array([2]));
	const aesKey = await crypto.subtle.importKey("raw", contentEncryptionKey, "AES-GCM", false, ["encrypt"]);
	const encrypted = new Uint8Array(await crypto.subtle.encrypt({
		name: "AES-GCM",
		iv: nonce
	}, aesKey, paddedPayload));
	const recordSize = new Uint8Array(4);
	new DataView(recordSize.buffer).setUint32(0, 4096);
	return concat(concat(salt, recordSize, new Uint8Array([localPublicKeyRaw.length]), localPublicKeyRaw), encrypted);
}
//#endregion
export { encryptPayload };
