// web-push-neo/dist/vapid.mjs
import { base64UrlToUint8Array, uint8ArrayToBase64Url, validateBase64Url } from "./base64url.mjs";
import { SignJWT, importPKCS8 } from "jose";
//#region src/vapid.ts
const DEFAULT_EXPIRATION_SECONDS = 720 * 60;
const RAW_PRIVATE_KEY_LENGTH = 32;
function importPrivateKey(decodedPrivateKey, decodedPublicKey) {
	if (decodedPrivateKey.length === RAW_PRIVATE_KEY_LENGTH) {
		const x = uint8ArrayToBase64Url(decodedPublicKey.slice(1, 33));
		const y = uint8ArrayToBase64Url(decodedPublicKey.slice(33, 65));
		const d = uint8ArrayToBase64Url(decodedPrivateKey);
		return crypto.subtle.importKey("jwk", {
			kty: "EC",
			crv: "P-256",
			x,
			y,
			d
		}, {
			name: "ECDSA",
			namedCurve: "P-256"
		}, false, ["sign"]);
	}
	return importPKCS8(`-----BEGIN PRIVATE KEY-----\n${btoa(String.fromCharCode(...decodedPrivateKey))}\n-----END PRIVATE KEY-----`, "ES256");
}
async function generateVAPIDKeys() {
	const keyPair = await crypto.subtle.generateKey({
		name: "ECDSA",
		namedCurve: "P-256"
	}, true, ["sign"]);
	const publicKeyRaw = new Uint8Array(await crypto.subtle.exportKey("raw", keyPair.publicKey));
	const privateKeyPkcs8 = new Uint8Array(await crypto.subtle.exportKey("pkcs8", keyPair.privateKey));
	return {
		publicKey: uint8ArrayToBase64Url(publicKeyRaw),
		privateKey: uint8ArrayToBase64Url(privateKeyPkcs8)
	};
}
function validateSubject(subject) {
	if (typeof subject !== "string" || subject.length === 0) throw new Error("The subject value must be a string containing an https: URL or mailto: address.");
	let parsed;
	try {
		parsed = new URL(subject);
	} catch {
		throw new Error(`Vapid subject is not a valid URL. ${subject}`);
	}
	if (parsed.protocol !== "https:" && parsed.protocol !== "mailto:") throw new Error(`Vapid subject is not an https: or mailto: URL. ${subject}`);
}
function validatePublicKey(publicKey) {
	if (typeof publicKey !== "string" || publicKey.length === 0) throw new Error("Vapid public key must be a URL-safe Base64 encoded string.");
	if (!validateBase64Url(publicKey)) throw new Error("Vapid public key must be URL-safe Base64 (without \"=\").");
	if (base64UrlToUint8Array(publicKey).length !== 65) throw new Error("Vapid public key should be 65 bytes long when decoded.");
}
function validatePrivateKey(privateKey) {
	if (typeof privateKey !== "string" || privateKey.length === 0) throw new Error("Vapid private key must be a URL-safe Base64 encoded string.");
	if (!validateBase64Url(privateKey)) throw new Error("Vapid private key must be URL-safe Base64 (without \"=\").");
}
async function createVapidAuthHeader(endpoint, publicKey, privateKey, subject, expiration) {
	const url = new URL(endpoint);
	const audience = `${url.protocol}//${url.host}`;
	const key = await importPrivateKey(base64UrlToUint8Array(privateKey), base64UrlToUint8Array(publicKey));
	const exp = expiration ?? Math.floor(Date.now() / 1e3) + DEFAULT_EXPIRATION_SECONDS;
	return `vapid t=${await new SignJWT({
		aud: audience,
		sub: subject
	}).setProtectedHeader({ alg: "ES256" }).setExpirationTime(exp).sign(key)},k=${publicKey}`;
}
//#endregion
export { createVapidAuthHeader, generateVAPIDKeys, validatePrivateKey, validatePublicKey, validateSubject };
