// web-push-neo/dist/web-push.mjs
import { base64UrlToUint8Array, validateBase64Url } from "./base64url.mjs";
import { Urgency } from "./constants.mjs";
import { encryptPayload } from "./encryption.mjs";
import { WebPushError } from "./error.mjs";
import { createVapidAuthHeader, validatePrivateKey, validatePublicKey, validateSubject } from "./vapid.mjs";
//#region src/web-push.ts
const URGENCY_VALUES = new Set(Object.values(Urgency));
function validateUrgency(urgency) {
	return URGENCY_VALUES.has(urgency);
}
function validateSubscription(subscription, hasPayload) {
	if (typeof subscription?.endpoint !== "string" || subscription.endpoint.length === 0) throw new Error("You must pass in a subscription with at least an endpoint.");
	if (!hasPayload) return;
	if (typeof subscription.keys?.p256dh !== "string" || typeof subscription.keys?.auth !== "string" || subscription.keys.p256dh.length === 0 || subscription.keys.auth.length === 0) throw new Error("To send a message with a payload, the subscription must have 'auth' and 'p256dh' keys.");
	if (base64UrlToUint8Array(subscription.keys.p256dh).length !== 65) throw new Error("The subscription p256dh value should be 65 bytes long.");
	if (base64UrlToUint8Array(subscription.keys.auth).length < 16) throw new Error("The subscription auth key should be at least 16 bytes long.");
}
function validateOptions(options) {
	const ttl = options?.TTL ?? 2419200;
	if (typeof ttl !== "number" || !Number.isInteger(ttl) || ttl < 0) throw new Error("TTL should be a non-negative integer.");
	const urgency = options?.urgency ?? Urgency.NORMAL;
	if (!validateUrgency(urgency)) throw new Error("Unsupported urgency specified.");
	const topic = options?.topic;
	if (topic !== void 0 && !validateBase64Url(topic)) throw new Error("Topic must use URL or filename-safe Base64 characters.");
	if (topic !== void 0 && topic.length > 32) throw new Error("Topic must be maximum 32 characters.");
	return {
		ttl,
		urgency,
		topic
	};
}
async function generateRequestDetails(subscription, payload, options) {
	const hasPayload = payload != null;
	validateSubscription(subscription, hasPayload);
	const { ttl, urgency, topic } = validateOptions(options);
	const headers = {
		TTL: String(ttl),
		Urgency: urgency,
		...options?.headers
	};
	if (topic !== void 0) headers["Topic"] = topic;
	let body = null;
	if (hasPayload) {
		body = await encryptPayload(payload, base64UrlToUint8Array(subscription.keys.p256dh), base64UrlToUint8Array(subscription.keys.auth));
		headers["Content-Length"] = String(body.length);
		headers["Content-Type"] = "application/octet-stream";
		headers["Content-Encoding"] = "aes128gcm";
	} else headers["Content-Length"] = "0";
	if (options?.vapidDetails !== void 0) {
		const { subject, publicKey, privateKey } = options.vapidDetails;
		validateSubject(subject);
		validatePublicKey(publicKey);
		validatePrivateKey(privateKey);
		headers["Authorization"] = await createVapidAuthHeader(subscription.endpoint, publicKey, privateKey, subject);
	}
	return {
		endpoint: subscription.endpoint,
		method: "POST",
		headers,
		body
	};
}
async function sendNotification(subscription, payload, options) {
	const requestDetails = await generateRequestDetails(subscription, payload, options);
	const response = await fetch(requestDetails.endpoint, {
		method: requestDetails.method,
		headers: requestDetails.headers,
		body: requestDetails.body,
		signal: options?.signal
	});
	const responseBody = await response.text();
	if (!response.ok) throw new WebPushError("Received unexpected response code", response.status, response.headers, responseBody, requestDetails.endpoint);
	return {
		statusCode: response.status,
		headers: response.headers,
		body: responseBody
	};
}
//#endregion
export { generateRequestDetails, sendNotification };
